File: src/server/src/main/java/com/project/domain/BaseEntity.java
package com.project.domain;

import com.project.utilities.Time;
import jakarta.persistence.MappedSuperclass;

import java.time.ZonedDateTime;

@MappedSuperclass
public abstract class BaseEntity {
    private final ZonedDateTime createdAt;
    private ZonedDateTime modifiedAt;
    private boolean isDeleted;

    protected BaseEntity() {
        this.createdAt = Time.utcNow();
    }

    public ZonedDateTime getCreatedAt() {
        return createdAt;
    }

    public ZonedDateTime getModifiedAt() {
        return modifiedAt;
    }

    public void setUpdatedAt() {
        this.modifiedAt = Time.utcNow();
    }

    public void setModifiedAt(ZonedDateTime modifiedAt) {
        this.modifiedAt = modifiedAt;
    }

    public void setIsDeleted(boolean deleted) {
        this.isDeleted = deleted;
    }

    public boolean getIsDeleted() {
        return this.isDeleted;
    }
}

File: src/server/src/main/kotlin/com/project/common/BaseEntity.kt
package com.project.common

import com.project.utilities.Time.utcNow
import jakarta.persistence.MappedSuperclass
import java.time.Instant

@MappedSuperclass
abstract class BaseEntity protected constructor() {
    var createdAt: Instant = utcNow()
    var modifiedAt: Instant? = null
    var isDeleted: Boolean = false

    fun markAsUpdated() {
        this.modifiedAt = utcNow()
    }

    fun markAsDeleted() {
        this.isDeleted = true
        this.markAsUpdated()
    }
}
